consensus/src/detail/tests/mod.rs

```diff
@@ -0,0 +1,221 @@
+            .collect::<Vec<_>>()
+    };
+
+    let outputs = {
+        let output_count = 1 + (rng.next_u32() as usize) % 10;
+        (0..output_count)
+            .into_iter()
+            .map(|_| generate_random_invalid_output(rng))
+            .collect::<Vec<_>>()
+    };
+
+    let flags = rng.next_u32();
+    let lock_time = rng.next_u32();
+
+    Transaction::new(flags, inputs, outputs, lock_time).expect(ERR_CREATE_TX_FAIL)
+}
+
+fn generate_random_invalid_block() -> Block {
+    let mut rng = rand::rngs::StdRng::from_entropy();
+
+    let transactions = {
+        let transaction_count = rng.next_u32() % 2000;
+        (0..transaction_count)
+            .into_iter()
+            .map(|_| generate_random_invalid_transaction(&mut rng))
+            .collect::<Vec<_>>()
+    };
+    let time = rng.next_u32();
+    let prev_id = Some(Id::new(&generate_random_h256(&mut rng)));
+
+    Block::new(transactions, prev_id, time, ConsensusData::None).expect(ERR_CREATE_BLOCK_FAIL)
+}
+
+fn create_utxo_data(
+    config: &ChainConfig,
+    tx_id: &Id<Transaction>,
+    index: usize,
+    output: &TxOutput,
+) -> Option<(TxInput, TxOutput)> {
+    if output.get_value() > Amount::from_atoms(1) {
+        Some((
+            TxInput::new(
+                OutPointSourceId::Transaction(tx_id.clone()),
+                index as u32,
+                random_witness(),
+            ),
+            TxOutput::new(
+                (output.get_value() - Amount::from_atoms(1)).unwrap(),
+                random_address(config),
+            ),
+        ))
+    } else {
+        None
+    }
+}
+
+fn produce_test_block(config: &ChainConfig, prev_block: &Block, orphan: bool) -> Block {
+    // For each output we create a new input and output that will placed into a new block.
+    // If value of original output is less than 1 then output will disappear in a new block.
+    // Otherwise, value will be decreasing for 1.
+    let (inputs, outputs): (Vec<TxInput>, Vec<TxOutput>) = prev_block
+        .transactions()
+        .iter()
+        .flat_map(|tx| create_new_outputs(config, tx))
+        .unzip();
+
+    Block::new(
+        vec![Transaction::new(0, inputs, outputs, 0).expect(ERR_CREATE_TX_FAIL)],
+        if orphan {
+            Some(Id::new(&H256::random()))
+        } else {
+            Some(Id::new(&prev_block.get_id().get()))
+        },
+        time::get() as u32,
+        ConsensusData::None,
+    )
+    .expect(ERR_CREATE_BLOCK_FAIL)
+}
+
+fn create_new_outputs(config: &ChainConfig, tx: &Transaction) -> Vec<(TxInput, TxOutput)> {
+    tx.get_outputs()
+        .iter()
+        .enumerate()
+        .filter_map(move |(index, output)| create_utxo_data(config, &tx.get_id(), index, output))
+        .collect::<Vec<(TxInput, TxOutput)>>()
+}
+
+fn wait_for_threadpool(consensus: &mut Consensus) {
+    // We continue execution when previous threads finished, or cause panic in 3 secs
+    let handle = consensus.events_broadcaster.spawn_handle(|| {});
```

Comment:
I don't understand this. A thread pool should have a queue to handle incoming events. Why is this necessary?

Reply:
Alrighty. If I understand this, you push this to wait for everything to finish in the queue? That's OK. Just please call the function `wait_for_threadpool_to_finish()`

Reply:
Yep, renamed